akka-persistence/src/main/scala/akka/persistence/temp/FSM.scala

```diff
@@ -0,0 +1,1157 @@
+      ref = Some(
+        if (repeat) scheduler.schedule(timeout, timeout, actor, this)
+        else scheduler.scheduleOnce(timeout, actor, this))
+
+    def cancel(): Unit =
+      if (ref.isDefined) {
+        ref.get.cancel()
+        ref = None
+      }
+  }
+
+  /**
+   * This extractor is just convenience for matching a (S, S) pair, including a
+   * reminder what the new state is.
+   */
+  object -> {
+    def unapply[S](in: (S, S)) = Some(in)
+  }
+
+  /**
+   * Log Entry of the [[akka.actor.LoggingFSM]], can be obtained by calling `getLog`.
+   */
+  final case class LogEntry[S, D](stateName: S, stateData: D, event: Any)
+
+  /**
+   * This captures all of the managed state of the [[akka.actor.FSM]]: the state
+   * name, the state data, possibly custom timeout, stop reason and replies
+   * accumulated while processing the last message.
+   */
+  final case class State[S, D, E](stateName: S, stateData: D, timeout: Option[FiniteDuration] = None, stopReason: Option[Reason] = None, replies: List[Any] = Nil, domainEvent: Option[E] = None)(private[akka] val notifies: Boolean = true) {
+
+    /**
+     * Copy object and update values if needed.
+     */
+    private[akka] def copy(stateName: S = stateName, stateData: D = stateData, timeout: Option[FiniteDuration] = timeout, stopReason: Option[Reason] = stopReason, replies: List[Any] = replies, notifies: Boolean = notifies, domainEvent: Option[E] = domainEvent): State[S, D, E] = {
+      State(stateName, stateData, timeout, stopReason, replies, domainEvent)(notifies)
+    }
+
+    /**
+     * Modify state transition descriptor to include a state timeout for the
+     * next state. This timeout overrides any default timeout set for the next
+     * state.
+     *
+     * Use Duration.Inf to deactivate an existing timeout.
+     */
+    def forMax(timeout: Duration): State[S, D, E] = timeout match {
+      case f: FiniteDuration ⇒ copy(timeout = Some(f))
+      case _                 ⇒ copy(timeout = None)
+    }
+
+    /**
+     * Send reply to sender of the current message, if available.
+     *
+     * @return this state transition descriptor
+     */
+    def replying(replyValue: Any): State[S, D, E] = {
+      copy(replies = replyValue :: replies)
+    }
+
+    /**
+     * Modify state transition descriptor with new state data. The data will be
+     * set when transitioning to the new state.
+     */
+    private[akka] def using(@deprecatedName('nextStateDate) nextStateData: D): State[S, D, E] = {
+      copy(stateData = nextStateData)
+    }
+
+    /**
+     * INTERNAL API.
+     */
+    private[akka] def withStopReason(reason: Reason): State[S, D, E] = {
+      copy(stopReason = Some(reason))
+    }
+
+    private[akka] def withNotification(notifies: Boolean): State[S, D, E] = {
+      copy(notifies = notifies)
+    }
+
+    /**
+     * Specify domain event that will to be applied when transitioning to the new state.
+     */
+    def applying(domainEvent: E): State[S, D, E] = {
```

Comment:
I would also like to one variant that takes a callback block as parameter, similar to `persist`. This block should be invoked after the event has been successfully been persisted (and after calling applyState/applyEvent).
That can be used for side effects, e.g. sending a message to someone else.